// messages/index.js
import forEach from 'lodash/forEach'
import isString from 'lodash/isString'
import frLocaleData from 'react-intl/locale-data/fr'
import React, {
  PropTypes
} from 'react'
import {
  addLocaleData,
  FormattedMessage,
  IntlProvider as IntlProvider_
} from 'react-intl'

// ===================================================================

const messages = {
  aboutPage: {
    defaultMessage: 'About'
  },
  homePage: {
    defaultMessage: 'Home'
  },

  usernameLabel: {
    defaultMessage: 'Username:'
  },
  passwordLabel: {
    defaultMessage: 'Password:'
  },
  signInButton: {
    defaultMessage: 'Sign in'
  }
}
forEach(messages, (message, id) => {
  if (isString(message)) {
    messages[id] = {
      id,
      defaultMessage: message
    }
  } else if (!message.id) {
    message.id = id
  }
})

addLocaleData(frLocaleData)
const frMessages = {
  aboutPage: 'À propos',
  homePage: 'Accueil',
  usernameLabel: 'Nom :',
  passwordLabel: 'Mot de passe :',
  signInButton: 'Connexion'
}

// ===================================================================

export default (messageId, values = {}) => <FormattedMessage
 {...messages[messageId]}
 values={values}
/>

export const IntlProvider = ({ children }) => (
  <IntlProvider_ locale='fr' messages={frMessages}>
    {children}
  </IntlProvider_>
)
IntlProvider.propTypes = PropTypes.node.isRequired
