// messages/index.js
import forEach from 'lodash/forEach'
import isString from 'lodash/isString'
import frLocaleData from 'react-intl/locale-data/fr'
import React, {
  Component,
  PropTypes
} from 'react'
import {
  connect
} from 'react-redux'
import {
  addLocaleData,
  FormattedMessage,
  IntlProvider as IntlProvider_
} from 'react-intl'

// ===================================================================

const messages = {
  // ----- Titles -----
  aboutPage: {
    defaultMessage: 'About'
  },
  homePage: {
    defaultMessage: 'Home'
  },

  // ----- Sign in -----
  usernameLabel: {
    defaultMessage: 'Username:'
  },
  passwordLabel: {
    defaultMessage: 'Password:'
  },
  signInButton: {
    defaultMessage: 'Sign in'
  },

  // ----- VM actions ------
  startVmLabel: {
    defaultMessage: 'Start VM'
  },
  recoveryModeLabel: {
    defaultMessage: 'Start VM in recovery mode'
  },
  suspendVmLabel: {
    defaultMessage: 'Suspend VM'
  },
  stopVmLabel: {
    defaultMessage: 'Stop VM'
  },
  forceShutdownVmLabel: {
    defaultMessage: 'Force shutdown VM'
  },
  rebootVmLabel: {
    defaultMessage: 'Reboot VM'
  },
  forceRebootVmLabel: {
    defaultMessage: 'Force reboot VM'
  },
  deleteVmLabel: {
    defaultMessage: 'Delete VM'
  },
  migrateVmLabel: {
    defaultMessage: 'Migrate VM'
  },
  snapshotVmLabel: {
    defaultMessage: 'Snapshot VM'
  },
  exportVmLabel: {
    defaultMessage: 'Export VM'
  },
  copyVmLabel: {
    defaultMessage: 'Copy VM'
  },
  cloneVmLabel: {
    defaultMessage: 'Clone VM'
  },
  convertToTemplateLabel: {
    defaultMessage: 'Convert VM to template'
  },
  vmConsoleLabel: {
    defaultMessage: 'VM console'
  },

  // ----- VM tabs -----
  generalTabName: {
    defaultMessage: 'General'
  },
  statsTabName: {
    defaultMessage: 'Stats'
  },
  consoleTabName: {
    defaultMessage: 'Console'
  },
  snapshotsTabName: {
    defaultMessage: 'Snapshots'
  },
  logsTabName: {
    defaultMessage: 'Logs'
  },
  advancedTabName: {
    defaultMessage: 'Advanced'
  },
  networkTabName: {
    defaultMessage: 'Network'
  },
  disksTabName: {
    defaultMessage: 'Disk{disks, plural, one {} other {s}}'
  },

  powerStateHalted: {
    defaultMessage: 'halted'
  },
  powerStateRunning: {
    defaultMessage: 'running'
  },

  // ----- VM general tab -----
  noToolsDetected: {
    defaultMessage: 'No Xen tools detected'
  },
  noIpv4Record: {
    defaultMessage: 'No IPv4 record'
  },
  noIpRecord: {
    defaultMessage: 'No IP record'
  },
  started: {
    defaultMessage: 'Started'
  },
  paraVirtualizedMode: {
    defaultMessage: 'Paravirtualization (PV)'
  },
  hardwareVirtualizedMode: {
    defaultMessage: 'Hardware virtualization (HVM)'
  },

  // ----- VM console tab -----
  copyToClipboardLabel: {
    defaultMessage: 'Copy'
  },
  ctrlAltDelButtonLabel: {
    defaultMessage: 'Ctrl+Alt+Del'
  },
  tipLabel: {
    defaultMessage: 'Tip:'
  },
  tipConsoleLabel: {
    defaultMessage: 'non-US keyboard could have issues with console: switch your own layout to US.'
  },

  // ----- VM network tab -----
  vifCreateDeviceButton: {
    defaultMessage: 'Create a new device'
  },
  vifNoInterface: {
    defaultMessage: 'No interface'
  },
  vifDeviceLabel: {
    defaultMessage: 'Device'
  },
  vifMacLabel: {
    defaultMessage: 'MAC address'
  },
  vifMtuLabel: {
    defaultMessage: 'MTU'
  },
  vifNetworkLabel: {
    defaultMessage: 'Network'
  },
  vifStatusLabel: {
    defaultMessage: 'Status'
  },
  vifStatusConnected: {
    defaultMessage: 'Connected'
  },
  vifStatusDisconnected: {
    defaultMessage: 'Disconnected'
  },
  vifIpAddresses: {
    defaultMessage: 'IP addresses'
  },

  // ----- VM snapshot tab -----
  noSnapshot: {
    defaultMessage: 'No snapshot'
  },
  tipCreateSnapshotLabel: {
    defaultMessage: 'Just click on the snapshot button to create one!'
  },
  snapshotDate: {
    defaultMessage: 'Creation date'
  },
  snapshotName: {
    defaultMessage: 'Name'
  },
  snapshotAction: {
    defaultMessage: 'Action'
  },

  // ----- VM advanced tab -----
  uuid: {
    defaultMessage: 'UUID'
  },
  virtualizationMode: {
    defaultMessage: 'Virtualization mode'
  },
  cpuWeightLabel: {
    defaultMessage: 'CPU weight'
  },
  defaultCpuWeight: {
    defaultMessage: 'Unspecified'
  },
  pvArgsLabel: {
    defaultMessage: 'PV args'
  },
  xenToolsStatus: {
    defaultMessage: 'Xen tools status'
  },
  xenToolsStatusValue: {
    defaultMessage: '{status}',
    description: 'status can be `not-installed`, `unknown`, `out-of-date` & `up-to-date`'
  },
  osName: {
    defaultMessage: 'OS name'
  },
  osKernel: {
    defaultMessage: 'OS kernel'
  },
  autoPowerOn: {
    defaultMessage: 'Auto power on'
  },
  ha: {
    defaultMessage: 'HA'
  },
  originalTemplate: {
    defaultMessage: 'Original template'
  },
  unknownOsName: {
    defaultMessage: 'Unknown'
  },
  unknownOsKernel: {
    defaultMessage: 'Unknown'
  },
  enabledAutoPowerOn: {
    defaultMessage: 'Enabled'
  },
  disabledAutoPowerOn: {
    defaultMessage: 'Disabled'
  },
  enabledHa: {
    defaultMessage: 'Enabled'
  },
  disabledHa: {
    defaultMessage: 'Disabled'
  },
  unknownOriginalTemplate: {
    defaultMessage: 'Unknown'
  }
}
forEach(messages, (message, id) => {
  if (isString(message)) {
    messages[id] = {
      id,
      defaultMessage: message
    }
  } else if (!message.id) {
    message.id = id
  }
})
const localizedMessages = {}

addLocaleData(frLocaleData)
localizedMessages.fr = {
  aboutPage: 'À propos',
  homePage: 'Accueil',
  usernameLabel: 'Nom :',
  passwordLabel: 'Mot de passe :',
  signInButton: 'Connexion',
  startVmLabel: 'Démarrer la VM',
  recoveryModeLabel: 'Démarrer la VM en mode sans échec',
  suspendVmLabel: 'Suspendre la VM',
  stopVmLabel: 'Arrêter la VM',
  forceShutdownVmLabel: 'Forcer l\'arrêt de la VM',
  rebootVmLabel: 'Redémarrer la VM',
  forceRebootVmLabel: 'Forcer le redémarrage de la VM',
  deleteVmLabel: 'Supprimer la VM',
  migrateVmLabel: 'Migrer la VM',
  snapshotVmLabel: 'Prendre un instantané de la VM',
  exportVmLabel: 'Exporter la VM',
  copyVmLabel: 'Copier la VM',
  cloneVmLabel: 'Cloner la VM',
  convertToTemplateLabel: 'Convertir la VM en modèle',
  vmConsoleLabel: 'Console de la VM',
  generalTabName: 'Général',
  statsTabName: 'Stats',
  consoleTabName: 'Console',
  snapshotsTabName: 'Instantanés',
  logsTabName: 'Journaux',
  advancedTabName: 'Avancé',
  networkTabName: 'Réseau',
  disksTabName: 'Disque{disks, plural, zero {} one {} other {s}}',
  powerStateHalted: 'arrêtée',
  powerStateRunning: 'en marche',
  started: 'Démarrée',
  noToolsDetected: 'Pas d\'outils Xen détectés',
  noIpv4Record: 'Aucune IPv4',
  noIpRecord: 'Aucune IP',
  virtualizationMode: 'Mode de virtualisation',
  // ----- VM console tab -----
  copyToClipboardLabel: 'Copier',
  ctrlAltDelButtonLabel: 'Ctrl+Alt+Suppr',
  tipLabel: 'Conseil :',
  tipConsoleLabel: 'Les agencements de clavier hors États-Unis ont des problèmes avec la console: passez le votre en "US".',
  // ----- VM advanced tab -----
  paraVirtualizedMode: 'Paravirtualisation (PV)',
  hardwareVirtualizedMode: 'Virtualisation matérielle (HVM)',
  xenToolsStatus: 'État des outils Xen',
  xenToolsStatusValue: `{status, select,
    unknown {Inconnu}
    up-to-date {À jour}
    out-of-date {Obsolètes}
    not-installed {Pas installés}
  }`,
  osName: 'Nom du système d\'exploitation',
  osKernel: 'Noyau du système d\'exploitation',
  autoPowerOn: 'Démarrage automatique',
  ha: 'Haute disponibilité',
  originalTemplate: 'Modèle d\'origine',
  unknownOsName: 'Inconnu',
  unknownOsKernel: 'Inconnu',
  enabledAutoPowerOn: 'Activé',
  disabledAutoPowerOn: 'Désactivé',
  enabledHa: 'Activée',
  disabledHa: 'Désactivée',
  unknownOriginalTemplate: 'Inconnu',
  // ----- VM network tab -----
  vifCreateDeviceButton: 'Créer une nouvelle interface',
  vifNoInterface: 'Aucune interface',
  vifDeviceLabel: 'Interface',
  vifMacLabel: 'Adresse physique',
  vifMtuLabel: 'MTU',
  vifNetworkLabel: 'Réseau',
  vifStatusLabel: 'Statut',
  vifStatusConnected: 'Connecté',
  vifStatusDisconnected: 'Déconnecté',
  vifIpAddresses: 'Adresses IP',
  // ----- VM network tab -----
  noSnapshot: 'Aucun instantané',
  tipCreateSnapshotLabel: 'Cliquez sur le bouton pour en créer un !',
  snapshotDate: 'Date de l\'instantané',
  snapshotName: 'Nom',
  snapshotAction: 'Action'
}

// ===================================================================

export default (messageId, values = {}) => <FormattedMessage
  {...messages[messageId]}
  values={values}
/>

@connect(({ lang }) => ({ lang }))
export class IntlProvider extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    lang: PropTypes.string.isRequired
  };

  render () {
    const { lang, children } = this.props
    return <IntlProvider_
      locale={lang}
      messages={localizedMessages[lang]}
    >
     {children}
    </IntlProvider_>
  }
}
