// xo-api/index.js
import angular from 'angular'
import indexOf from 'lodash.indexof'
const isArray = angular.isArray

import 'angular-cookies'

import xoLib from 'xo-lib'

export default angular.module('xo-api', [
  'ngCookies'
])
  .run(function ($rootScope) {
    // Ensure correct integration with Angular.
    xoLib.setScheduler(function (fn) {
      $rootScope.$evalAsync(fn)
    })
  })
  .service('xoApi', function ($cookieStore) {
    const xo = new xoLib.Xo()

    try {
      const token = $cookieStore.get('token')

      // If there is a token, sign in with it.
      if (token) {
        xo.signIn({ token })
      }
    } catch (e) {
      if (e instanceof SyntaxError) {
        $cookieStore.remove('token')
      } else {
        throw e
      }
    }

    const getObject = (function (objects) {
      const {
        all: byIds,
        indexes: {
          ref: byRefs,
          UUID: byUuids
        }
      } = objects

      return function getObject (id, type) {
        const object = byIds[id] || byUuids[id] || byRefs[id]

        if (
          // The object has been found and …
          object && (
            // … no type specified.
            !type ||

            // … is of the expected type.
            (type === object.type) ||

            // … is of one of the allowed types.
            isArray(type) && (indexOf(type, object.type) === -1)
          )
        ) {
          return object
        }
      }
    })(xo.objects)

    return {
      // -----------------
      // Session
      // -----------------

      logIn (email, password, persist) {
        return xo.signIn({ email, password }).then(() => {
          if (persist) {
            xo.call('token.create').then(function (token) {
              $cookieStore.put('token', token)
            })
          }
        })
      },
      logOut () {
        $cookieStore.remove('token')

        return xo.signOut()
      },
      get status () {
        return xo.status
      },
      get user () {
        return xo.user
      },

      // -----------------
      // RPC
      // -----------------

      call (method, params) {
        return xo.call(method, params)
      },

      // -----------------
      // Objects
      // -----------------

      get (id, types) {
        if (isArray(id)) {
          const objects = []

          angular.forEach(id, id => {
            const object = getObject(id, types)
            if (object) {
              objects.push(object)
            }
          })

          return objects
        }

        return getObject(id, types)
      },

      all: xo.objects.all,

      byTypes: xo.objects.indexes.type
    }
  })
  .name
