// xo-api/index.js
import angular from 'angular';
import 'angular-cookies';

import xoLib from 'xo-lib';

export default angular.module('xo-api', [
  'ngCookies',
])
  .run(function ($rootScope) {
    // Ensure correct integration with Angular.
    xoLib.setScheduler(function (fn) {
      $rootScope.$evalAsync(fn);
    });
  })
  .service('xoApi', function ($cookieStore) {
    var xo = new xoLib.Xo();

    try {
      let token = $cookieStore.get('token');

      // If there is a token, sign in with it.
      if (token) {
        xo.signIn({ token });
      }
    } catch (e) {
      if (e instanceof SyntaxError) {
        $cookieStore.remove('token');
      } else {
        throw e;
      }
    }

    return {
      //------------------
      // Session
      //------------------

      logIn(email, password, persist) {
        return xo.signIn({ email, password }).then(() => {
          if (persist) {
            xo.call('token.create').then(function (token) {
              $cookieStore.put('token', token);
            });
          }
        });
      },
      logOut() {
        $cookieStore.remove('token');

        return xo.signOut();
      },
      get status() {
        return xo.status;
      },
      get user() {
        return xo.user;
      },

      //------------------
      // RPC
      //------------------

      call(method, params) {
        return xo.call(method, params);
      },

      //------------------
      // Objects
      //------------------

      get(id) {
        let all = xo.objects.all;
        let byRefs = xo.objects.indexes.ref;
        let byUuids = xo.objects.indexes.UUID;

        if (angular.isArray(id)) {
          let objects = [];
          angular.forEach(id, function (id) {
            let object = all[id];

            if (!object) {
              object = byUuids[id] || byRefs[id];

              if (!object) {
                return;
              }

              object = object[0];
            }

            objects.push(object);
          });
          return objects;
        }

        let object = all[id];
        if (object) {
          return object;
        }

        object = byUuids[id] || byRefs[id];
        if (object) {
          return object[0];
        }
      },

      all: xo.objects.all,

      byTypes: xo.objects.indexes.type,
    };
  })
  .name
;
