// tag/index.js
import angular from 'angular'
import trim from 'lodash.trim'
import pull from 'lodash.pull'

import view from './view'

// =====================================================================

export default angular.module('xoWebApp.tag', [])

  .directive('xoTag', () => {
    function link (scope, element, attrs) {
      element.find('.add-button').on('click', function () {
        element.find('input').focus()
      })
    }

    return {
      restrict: 'E',
      template: view,
      scope: {
        object: '='
      },
      controller: 'XoTag as ctrl',
      bindToController: true,
      link
    }
  })

  .controller('XoTag', function ($scope, xo, xoApi) {
    const {get} = xoApi
    if (typeof this.object === 'string') {
      this.object = get(this.object)
    }

    $scope.canAdmin = (id) => {
      return id && xoApi.canInteract(id, 'administrate') || false
    }

    this.add = (tag) => {
      tag = trim(tag)
      if (tag === '') {
        return
      }
      xo.tag.add(tag, this.object.id)
      .then(() => this.object.tags.push(tag)) // FIXME Work around until refresh problems solved
    }

    this.remove = (tag) => {
      xo.tag.remove(tag, this.object.id)
      .then(() => pull(this.object.tags, tag)) // FIXME Work around until refresh problems solved
    }
  })

  // A module exports its name.
  .name
