// xo-parallel-d3/index.js
'use strict'

import angular from 'angular'
import clone from 'lodash.clonedeep'
import foreach from 'lodash.foreach'
import d3 from 'd3'

export default angular.module('xoParallelD3', [])
  .directive('parallelChart', function ($parse) {
    function link (scope, element, attrs) {
      var traits,
        svg,
        width,
        height,
        x, y,
        line, g,
        axis,
        color,
        text,
        chart_data

      width = attrs.width ? parseInt(attrs.width, 10) : 2000
      height = attrs.height ? parseInt(attrs.height, 10) : width / 2/* breadcrumbs */

      line = d3.svg.line()
      axis = d3.svg.axis().orient('left')
      /* don't want decimals */
        .tickFormat(d3.format('d'))
      color = d3.scale.category10()
       // took responsive svg trick from http://demosthenes.info/blog/744/Make-SVG-Responsive
      svg = d3.select(element[0])
        .append('div')
        /* padding from http://soqr.fr/testsvg/embed-svg-liquid-layout-responsive-web-design.php */
        .attr('style', 'position: relative;width: 90%;padding-bottom: 50%;margin-left:5%; vertical-align: middle;overflow: hidden;')
        .append('svg')
        .attr('style', 'display: block;position: absolute; top: 0;left: 0;')
        .attr('width', '100%')
        .attr('height', '100%')
        .attr('preserveAspectRatio', 'xMinYMin meet')
        .attr('viewBox', '0 0 ' + width + ' ' + height)
        .append('g')
        .attr('transform', 'translate(' + width / 20 + ',' + height / 10 + ')')
        .style('font-size', '30px')/* since viewbox is 2000px , font should be big */

      scope.$watch(() => scope.chartData, function (newVal) {
        // purge existing graph ( no animation )
        svg
          .selectAll('.trait')
          .remove()

        svg
          .selectAll('.axis')
          .remove()

        svg
          .selectAll('.line')
          .remove()
        y = {}
        traits = []
        if (!scope.chartData && scope.chartData.length > 0) {
          return
        }

      // d3 add a lot of data to nodes. It triggers watch without rest
        chart_data = clone(scope.chartData)

      // populate traits with the key of the first record
        foreach(chart_data[0], function (value, key) {
          if (key !== 'name' && key !== 'id') {
            traits.push(key)
          }
        })

        x = d3.scale.ordinal().domain(traits).rangePoints([0, width * 0.9])
        // Create a scale and brush for each trait.
        foreach(traits, function (trait, idx) {
          // Coerce values to numbers.
          chart_data.forEach(function (datum) { datum[idx] = +datum[idx] })

          y[trait] = d3.scale.linear()
              .domain(d3.extent(chart_data, function (datum) { return datum[trait] }))
              .range([height, 0])
        })

      // Add foreground lines.
        svg.append('svg:g')
          .attr('class', 'foreground')
          .style({stroke: 'red', fill: 'none', 'stroke-opacity': 0.5, 'stroke-width': width / 250})
          .selectAll('path')
          .data(chart_data)
          .enter().append('svg:path')
          .attr('d', path)
          .attr('class', 'line')
          .attr('id', function (d) {return 'line-' + d.id})
          .attr('stroke', (node) => color(node.name))
          .on('mouseover', mouseover)
          .on('mouseout', () => highlight())
          .on('click', click)
      //  .attr('class', function(d) { return d.species; });

      // Add a group element for each trait.
        g = svg.selectAll('.trait')
          .data(traits)
          .enter().append('svg:g')
          .attr('class', 'trait')
          .attr('transform', function (d) { return 'translate(' + x(d) + ')' })

      // Add an axis and title.
        g.append('svg:g')
          .attr('class', 'axis')
          .each(function (d) { d3.select(this).call(axis.scale(y[d])) })
          .append('svg:text')
          .style('font-size', '125%')
          .style('font-weight', 'bold')
          .attr('text-anchor', 'middle')
          .attr('y', -50)
          .text(String)
      })

      function mouseover (node) {
        highlight(node, d3.mouse(this))
        if (scope.over) {
          scope.over.apply(null, [{ d: node }])
        }
      }

      function click (node) {
        if (scope.click) {
          scope.click.apply(null, [{ d: node }])
        }
      }

      function highlight (node, position) {
        let node_id
        if (node) {
          node_id = node.id
        }
        if (text) {
          text.remove()
        }

        if (position) {
          let bbox, padding, t

          text = svg.append('svg:g')
          t = text.append('svg:text')
            .text(node.name)
            .attr('x', position[0])
            .attr('y', position[1] - 30)
            .style({'font-size': '125%', 'font-weight': 'bold', 'fill': 'white'})
          bbox = t.node().getBBox()
          padding = 10
          text.insert('svg:rect', '*')
            .attr('x', bbox.x - padding)
            .attr('y', bbox.y - padding)
            .attr('width', bbox.width + (padding * 2))
            .attr('height', bbox.height + (padding * 2))
            .style('fill', color(node.name))
        }

        svg
          .selectAll('.line')
          .style({'stroke-opacity': 0.5, 'stroke-width': width / 250})
        svg
          .select('#line-' + node_id)
          .style({'stroke-opacity': 1, 'stroke-width': width / 125})
      }
  // Returns the path for a given data point.
      function path (d) {
        return line(
            traits.map(
               function (trait) {
                  return [x(trait), y[trait](d[trait])]
                }
           )
        )
      }

    }

    return {
      resctict: 'E',
      replace: false,
      scope: {
        chartData: '=',
        selected: '=',
        log: '=',
        over: '&',
        click: '&'
      },
      link: link
    }

  })
  // A module exports its name.
  .name
