// xo-week-heatmap/index.js
'use strict'

import angular from 'angular'
import foreach from 'lodash.foreach'
import map from 'lodash.map'
import moment from 'moment'

import view from './view'

export default angular
  .module('xoWeekHeatmap', [])
  .directive('weekheatmap', heatmap)
  .name

function heatmap () {
  return {
    restrict: 'E',
    replace: false,
    scope: {
      chartData: '=',
      steps: '=?',
      colors: '=?'
    },
    link,
    template: view
  }

  function link (scope, element, attrs) {
    scope.matrix = []
    scope.colors = scope.colors || ['#ffffee', '#ffffd9', '#edf8b1', '#c7e9b4', '#7fcdbb', '#41b6c4', '#1d91c0', '#225ea8', '#253494', '#081d58']

    scope.hourLabels = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]

    if (scope.steps) {
      scope.intervals = []

      for (var i = 1; i < scope.steps.length; i++) {
        scope.intervals.push({
          start: scope.steps[i - 1],
          end: scope.steps[i],
          color: scope.colors[i]
        })
      }
    }
    scope.$watch(function () { return scope.chartData }, function (newVal) {
      displayData()
    })

    function displayData () {
      if (!scope.chartData) {
        return
      }
      const chartData = scope.chartData.values && scope.chartData || {values: scope.chartData}

      scope.min = Math.min.apply(null, map(chartData.values, 'value'))
      scope.max = Math.max.apply(null, map(chartData.values, 'value'))
      if (!scope.steps) {
        scope.intervals = []
        foreach(scope.colors, function (color, idx) {
          scope.intervals.push({
            start: scope.min + idx * (scope.max - scope.min) / scope.colors.length,
            end: scope.min + (1 + idx) * (scope.max - scope.min) / scope.colors.length,
            color: color
          })
        })
      }
      var matrix = {}
      let matrixLength = 0
      let dayLength = 0
      let uniqueId = 0
      foreach(chartData.values, function (datum) {
        const d = new Date(datum.date)
        const v = datum.value
        const day = moment(d).format('Do')
        const hour = d.getHours()
        if (!matrix[day]) {
          matrix[day] = []
          matrixLength++
        }
        if (!matrix[day][hour]) {
          matrix[day][hour] = {
            value: v,
            nb: 1,
            id: uniqueId++,
            dateTip: moment(d).format('ddd DD MMM'),
            filter: chartData.filter || (identity => identity)
          }
          dayLength = Math.max(dayLength, hour)
        } else {
          matrix[day][hour].value += v
          matrix[day][hour].nb++
          matrix[day][hour].dateTip = moment(d).format('ddd DD MMM')
        }
        matrix[day][hour].color = color(matrix[day][hour].value / matrix[day][hour].nb)
      })
      foreach(matrix, day => {
        for (let pos = 0; pos <= dayLength; pos++) {
          if (!day[pos]) {
            day[pos] = {
              value: null,
              nb: 1,
              id: uniqueId++,
              color: '#ffffff'
            }
          }
        }
      })

      scope.matrix = matrix
      scope.matrixLength = matrixLength
    }

    function color (value) {
      var color = null
      foreach(scope.intervals, function (interval, pos) {
        if (interval.start <= value && value < interval.end || (pos === scope.intervals.length - 1 && value === interval.end)) {
          color = interval.color
        }
      })
      return color
    }

    displayData()
  }
}
