// angular-no-vnc/index.js
import angular from 'angular';
import {parse as parseUrl, resolve as resolveUrl} from 'url';
import {RFB} from 'novnc-node';

import view from './view';

function parseRelativeUrl(url) {
  /* global window: false */
  return parseUrl(resolveUrl(String(window.location), url));
}

export default angular.module('no-vnc', [])
  .controller('NoVncCtrl', function ($attrs, $element, $scope) {
    this.height = 480;
    $attrs.$observe('height', (height) => {
      this.height = height;
    });
    this.width = 640;
    $attrs.$observe('width', (width) => {
      this.width = width;
    });

    let rfb;
    function clean() {
      // If there was a previous connection.
      if (rfb) {
        rfb.disconnect();
        rfb = undefined;
      }
    }

    this.remoteControl = {
      sendCtrlAltDel() {
        if (rfb) {
          rfb.sendCtrlAltDel();
        }
      },
    };

    let canvas = $element.find('canvas')[0];

    $attrs.$observe('url', (url) => {
      // Remove previous connection.
      clean();

      // If the URL is empty, stop now.
      if (!url) {
        return;
      }

      // Parse the URL.
      url = parseRelativeUrl(url);
      let isSecure =
        (url.protocol === 'https:') ||
        (url.protocol === 'wss:')
      ;

      rfb = new RFB({
        encrypt: isSecure,
        target: canvas,
        wsProtocols: ['chat'],
      });

      // Leading '/' is already added by noVNC.
      let path = url.path;
      if (path[0] === '/') {
        path = path.substr(1);
      }

      let port = url.port;
      if (!port) {
        port = isSecure ? 443 : 80;
      }

      // Connect.
      rfb.connect(
        url.hostname,
        port,
        '', // TODO: comment.
        path
      );
    });

    $scope.$on('$destroy', clean);
  })
  .directive('noVnc', function () {
    return {
      bindToController: true,
      controller: 'NoVncCtrl as noVnc',
      restrict: 'E',
      scope: {
        remoteControl: '=',
      },
      template: view,
    };
  })
  .name
;
