// xo-sparklines/index.js
import map from 'lodash/map'
import React from 'react'
import {
  Sparklines,
  SparklinesLine,
  SparklinesReferenceLine,
  SparklinesSpots
} from 'react-sparklines'
import { propTypes } from 'utils'

const STYLE = {}

const WIDTH = 100
const HEIGHT = 30

// ===================================================================

function computeArraysAvg (arrays) {
  if (!arrays || !arrays.length || !arrays[0].length) {
    return
  }

  const n = arrays[0].length
  const m = arrays.length

  const result = new Array(n)

  for (let i = 0; i < n; i++) {
    result[i] = 0

    for (let j = 0; j < m; j++) {
      result[i] += arrays[j][i]
    }

    result[i] / m
  }

  return result
}

function computeObjectsAvg (objects) {
  return computeArraysAvg(
    map(objects, (object) =>
      computeArraysAvg(map(object, (arr) => arr))
    )
  )
}

// ===================================================================

export const CpuSparkLines = propTypes({
  data: propTypes.object.isRequired
})(({ data }) =>
  <Sparklines style={STYLE} data={computeArraysAvg(data)} max={100} width={WIDTH} height={HEIGHT}>
    <SparklinesLine color='blue'/>
    <SparklinesSpots />
  </Sparklines>
)

export const MemorySparkLines = propTypes({
  data: propTypes.object.isRequired
})(({ data }) => {
  const { memory } = data

  return (
    <Sparklines style={STYLE} data={data.memoryUsed} max={memory[memory.length - 1]} width={WIDTH} height={HEIGHT}>
      <SparklinesLine color='red' />
      <SparklinesReferenceLine type='mean' />
      <SparklinesSpots />
    </Sparklines>
  )
})

export const XvdSparkLines = propTypes({
  data: propTypes.object.isRequired
})(({ data }) =>
  <Sparklines style={STYLE} data={computeObjectsAvg(data)} width={WIDTH} height={HEIGHT}>
    <SparklinesLine color='purple' />
    <SparklinesSpots />
  </Sparklines>
)

export const VifSparkLines = propTypes({
  data: propTypes.object.isRequired
})(({ data }) =>
  <Sparklines style={STYLE} data={computeObjectsAvg(data)} width={WIDTH} height={HEIGHT}>
    <SparklinesLine color='green' />
    <SparklinesSpots />
  </Sparklines>
)
